tests/run_test.cmake

```diff
@@ -1,113 +1,91 @@
+
+  if(uncrustify_error_code)
+    message(SEND_ERROR
+      "Uncrustify error_code ${uncrustify_error_code} (${name})"
     )
+  else()
     execute_process(
-      COMMAND ${CMAKE_COMMAND} -E copy ${TEST_DIR}/${TEST_RESULT} /dev/stderr
+      COMMAND ${CMAKE_COMMAND} -E compare_files ${result} ${output}
+      RESULT_VARIABLE files_are_different
     )
-    message(SEND_ERROR "Uncrustify MISSMATCH (Pass 1)")
+    if(files_are_different)
+      message(WARNING
+        "MISMATCH (${name}): ${result} does not match ${output}"
+      )
+      if(GIT_EXECUTABLE)
+        execute_process(
+          COMMAND ${GIT_EXECUTABLE}
+            diff --no-index
+            ${output}
+            ${result}
+        )
+      else()
+        execute_process(
+          COMMAND ${CMAKE_COMMAND} -E echo "the result file:"
+        )
+        execute_process(
+          COMMAND ${CMAKE_COMMAND} -E copy ${result} /dev/stderr
+        )
+      endif()
+      message(SEND_ERROR "Uncrustify MISMATCH (${name})")
+    endif()
   endif()
-endif(uncrustify_error_code)
-file(REMOVE ${TEST_DIR}/ERR-1.txt)
+endfunction()
 
+require(
+  TEST_PROGRAM
+  TEST_LANG
+  TEST_CONFIG
+  TEST_RERUN_CONFIG
+  TEST_INPUT
+  TEST_OUTPUT
+  TEST_RESULT
+  TEST_RESULT_2
+  TEST_DIR
+  TEST_RERUN_OUTPUT
+)
 
-# Re-run with the output file as the input to check stability.
-execute_process(
-  COMMAND ${TEST_PROGRAM} -l ${TEST_LANG} -c ${TEST_RERUN_CONFIG} -f ${TEST_RERUN_OUTPUT} -o ${TEST_RESULT_2}
-    WORKING_DIRECTORY ${TEST_DIR}
-    RESULT_VARIABLE uncrustify_error_code
-    OUTPUT_QUIET
-    ERROR_VARIABLE uncrustify_error
-    ERROR_FILE ${TEST_DIR}/ERR-2.txt
+# Run first pass
+run_test(
+  "Pass 1"
+  ${TEST_CONFIG}
+  ${TEST_DIR}/${TEST_INPUT}
+  ${TEST_DIR}/${TEST_RESULT}
+  ${TEST_DIR}/${TEST_OUTPUT}
 )
 
-if(uncrustify_error_code)
-  execute_process(
-    COMMAND ${CMAKE_COMMAND} -E echo "Uncrustify error. The ERR-2.txt file is:"
-  )
-  execute_process(
-    COMMAND ${CMAKE_COMMAND} -E copy ${TEST_DIR}/ERR-2.txt /dev/stderr
-  )
-  message(SEND_ERROR "Uncrustify error_code (Pass 2)")
-else(uncrustify_error_code)
-  execute_process(
-    COMMAND ${CMAKE_COMMAND} -E compare_files ${TEST_DIR}/${TEST_RESULT_2} ${TEST_DIR}/${TEST_RERUN_OUTPUT}
-      RESULT_VARIABLE files_are_different
-  )
-  if(files_are_different)
-    message(WARNING "UNSTABLE (Pass 2): ${TEST_RESULT_2} does not match ${TEST_RERUN_OUTPUT}")
-    execute_process(
-      COMMAND ${CMAKE_COMMAND} -E echo "the result file:"
-    )
-    execute_process(
-      COMMAND ${CMAKE_COMMAND} -E copy ${TEST_DIR}/${TEST_RESULT_2} /dev/stderr
-    )
-    message(SEND_ERROR "Uncrustify UNSTABLE (Pass 2)")
-  endif()
-endif()
-file(REMOVE ${TEST_DIR}/ERR-2.txt)
+# Re-run with the output file as the input to check stability
+run_test(
+  "Pass 2"
+  ${TEST_RERUN_CONFIG}
+  ${TEST_DIR}/${TEST_RERUN_OUTPUT}
+  ${TEST_DIR}/${TEST_RESULT_2}
+  ${TEST_DIR}/${TEST_RERUN_OUTPUT}
```

Comment:
Are you sure about this? This doesn't make sense for me.

Reply:
Uh... I am sure that matches *the old behavior*, yes :stuck_out_tongue:. The old invocation was:
https://github.com/uncrustify/uncrustify/blob/21651401f80f1e833d3e38ead78dddb3156f06af/tests/run_test.cmake#L81
...which is indeed passing `${TEST_RERUN_OUTPUT}` as the input.

It sure seems like the input is supposed to be `TEST_RESULT` (or `TEST_OUTPUT`), though, doesn't it? (Of course, if the test is *working*, and isn't using the special re-run logic, they're all the same, which is probably why no one noticed before...)

Please let me know which is correct :slightly_smiling_face:.

Reply:
No that is actually correct, we have three dirs: input, output (expected), result (generated).

Reply:
yes, it's very confusing otherwise.

Actually @CDanU noticed they were not working https://github.com/uncrustify/uncrustify/issues/907#issuecomment-394465167 so this might be a good chance to fix them here.

I later want to turn instabilities as errors #907 but we have to make sure they are still being reported as warnings now.

Reply:
Since I am now confused :slightly_smiling_face:, I made the other requested change. AFAICT the behavior as of the current state of this PR is equivalent to the old behavior. (Note: both passes eventually do/did a `SEND_ERROR` if the output and expected differ. The second pass is "interesting" in that it uses the expected output as the input, but... hmm...)

See:
https://github.com/uncrustify/uncrustify/blob/21651401f80f1e833d3e38ead78dddb3156f06af/tests/run_test.cmake#L73
https://github.com/uncrustify/uncrustify/blob/21651401f80f1e833d3e38ead78dddb3156f06af/tests/run_test.cmake#L110

Re: #907... I notice that (for the second pass) `run_test.cmake` uses the expected second-pass output as input, while `run_tests.py` uses the expected *first* pass output as input. With the former approach, all tests are passing. With the latter, I see three failures on my machine: `c-sharp_60009`, `c_02102`, and `cpp_30265`. That said... it looks like `run_tests.py` doesn't use the rerun_expected *at all*?

Also concerning is that I see now that `run_test.py` runs some tests that are missing from CTest...

Reply:
> Also concerning is that I see now that run_test.py runs some tests that are missing from CTest...

I will make a separate PR to address this... I'm not sure what we'll want to do with it, though, since it will cause CI to start failing.

Reply:
I have some thoughts, but... later... (Meanwhile, I notice that this checks off 'diff output' in #977 :smile:.)